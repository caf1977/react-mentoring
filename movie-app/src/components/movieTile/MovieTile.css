.movie-tile {
    border: 1px solid;
    padding: 10px;
    cursor: pointer;
}

.movie-tile img {
    height: 350px;
    width: 300px;
}

.movie-tile p {
    margin: 0;
    font-size: 14px;
    color: #666;
}

.movie-tile-info {
    display: flex;
}

.movie-tile-info-year {
    padding-left: 70px;
}

.movie-tile-info-year h2 {
    border: 1px solid;
    margin: 2px;
    font-size: 15px;
    padding-left: 5px;
    padding-right: 5px;
}

.movie-tile-info h3 {
    margin: 0;
    font-size: 18px;
}
