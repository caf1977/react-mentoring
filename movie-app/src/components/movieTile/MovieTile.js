import { useState } from 'react';
import './MovieTile.css'
import { useNavigate, useSearchParams } from 'react-router';

const MovieTile = ({movieInfo, onClick}) => {
    const [searchParams] = useSearchParams();
    const [menuOpen, setMenuOpen] = useState(false);
    const navigate = useNavigate();

    const handleMenuToggle = (e) => {
        e.stopPropagation();
        setMenuOpen(!menuOpen);
    };

    const handleEdit = (e) => {
        e.stopPropagation();
        navigate(`/${movieInfo.id}/edit?${searchParams.toString()}`);
        setMenuOpen(false);
    };

    const handleDelete = (e) => {
        e.stopPropagation();
        setMenuOpen(false); 
    };

    return (
        <div data-testid="info" className="movie-tile" onClick={() => onClick(movieInfo)}>
            <div className="image-container">
                <img src={movieInfo.poster_path} alt={movieInfo.title}/>
                <button className="three-dots-button" onClick={handleMenuToggle}>
                    &#8230;
                </button>
                
                {menuOpen && (
                    <div className="menu">
                        <button onClick={handleEdit}>Edit</button>
                        <button onClick={handleDelete}>Delete</button>
                    </div>
                )}
            </div>
            <div className="movie-tile-info">
                <h3>{movieInfo.title}</h3>
                <div className="movie-tile-info-year">
                    <h2>
                        {movieInfo.release_date}
                    </h2>
                </div>
            </div>
            <p>{movieInfo.genres.join(",")}</p>
        </div>
    )
};

export default MovieTile;