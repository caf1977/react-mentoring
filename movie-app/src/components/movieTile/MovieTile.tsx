import React, { FC, useState, MouseEvent } from 'react';
import './MovieTile.css'
import { useNavigate, useSearchParams } from 'react-router-dom';
import { MovieInfo } from '../../types/MovieInfo';

interface IMovieTile {
    movieInfo: MovieInfo;
    onClick?: (movie: MovieInfo) =>void;
}

const MovieTile: FC<IMovieTile> = ({movieInfo, onClick}) => {
    const [searchParams] = useSearchParams();
    const [menuOpen, setMenuOpen] = useState(false);
    const navigate = useNavigate();

    const handleMenuToggle = (e: MouseEvent<HTMLButtonElement>) => {
        e.stopPropagation();
        setMenuOpen(!menuOpen);
    };

    const handleEdit = (e: MouseEvent<HTMLButtonElement>) => {
        e.stopPropagation();
        navigate(`/${movieInfo.id}/edit?${searchParams.toString()}`);
        setMenuOpen(false);
    };

    const handleDelete = (e: MouseEvent<HTMLButtonElement>) => {
        e.stopPropagation();
        setMenuOpen(false); 
    };

    return (
        <div data-testid="info" className="movie-tile" onClick={() => onClick?.(movieInfo)}>
            <div className="image-container">
                <img src={movieInfo.poster_path} alt={movieInfo.title}/>
                <button className="three-dots-button" onClick={handleMenuToggle}>
                    &#8230;
                </button>
                
                {menuOpen && (
                    <div className="menu">
                        <button onClick={handleEdit}>Edit</button>
                        <button onClick={handleDelete}>Delete</button>
                    </div>
                )}
            </div>
            <div className="movie-tile-info">
                <h3>{movieInfo.title}</h3>
                <div className="movie-tile-info-year">
                    <h2>
                        {movieInfo.release_date}
                    </h2>
                </div>
            </div>
            <p>{movieInfo.genres.join(",")}</p>
        </div>
    )
};

export default MovieTile;