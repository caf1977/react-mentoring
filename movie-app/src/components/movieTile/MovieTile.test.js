import {render, screen, fireEvent} from "@testing-library/react";
import MovieTile from "./MovieTile";

describe("MovieTile component tests", () => {

    const movie = {
        poster_path: "https://m.media-amazon.com/images/M/MV5BN2U4OTdmM2QtZTkxYy00ZmQyLTg2N2UtMDdmMGJmNDhlZDU1XkEyXkFqcGc@._V1_FMjpg_UX1000_.jpg",
        title: "Mission Impossible",
        release_date: 2018,
        genres: ["Drama", "Thriller"],
    };

    it("renders movie tile info", () => {
        render(<MovieTile movieInfo={movie}/>);
        const image = screen.getByAltText("Mission Impossible");
        const name = screen.getByText("Mission Impossible");
        const releaseYear = screen.getByText("2018");
        const genres = screen.getByText("Drama,Thriller");
        
        expect(image.src).toContain("https://m.media-amazon.com/images/M/MV5BN2U4OTdmM2QtZTkxYy00ZmQyLTg2N2UtMDdmMGJmNDhlZDU1XkEyXkFqcGc@._V1_FMjpg_UX1000_.jpg");
        expect(name).toBeInTheDocument();
        expect(releaseYear).toBeInTheDocument();
        expect(genres).toBeInTheDocument();
    })

    it("onClick is called when image is clicked", () => {
        window.alert = jest.fn();
        render(<MovieTile movieInfo={movie} onClick={(value) => {alert(value)}}/>);
        const tile = screen.getByTestId("info");
        fireEvent.click(tile);

        expect(window.alert).toHaveBeenCalledWith(movie);
    })

    it("renders snapshot", () => {
        const { asFragment } = render(<MovieTile movieInfo={movie}/>);

        expect(asFragment(<MovieTile movieInfo={movie}/>)).toMatchSnapshot();
    })
})