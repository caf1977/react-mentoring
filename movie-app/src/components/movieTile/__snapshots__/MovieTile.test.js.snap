// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie component tests render snapshot 1`] = `
<DocumentFragment>
  <div
    class="movie-tile"
    data-testid="info"
  >
    <img
      alt="Mission Impossible"
      src="https://m.media-amazon.com/images/M/MV5BN2U4OTdmM2QtZTkxYy00ZmQyLTg2N2UtMDdmMGJmNDhlZDU1XkEyXkFqcGc@._V1_FMjpg_UX1000_.jpg"
    />
    <div
      class="movie-tile-info"
    >
      <h3>
        Mission Impossible
      </h3>
      <div
        class="movie-tile-info-year"
      >
        <h2>
          2018
        </h2>
      </div>
    </div>
    <p>
      Drama,Thriller
    </p>
  </div>
</DocumentFragment>
`;

exports[`MovieTile component tests render snapshot 1`] = `
<DocumentFragment>
  <div
    class="movie-tile"
    data-testid="info"
  >
    <img
      alt="Mission Impossible"
      src="https://m.media-amazon.com/images/M/MV5BN2U4OTdmM2QtZTkxYy00ZmQyLTg2N2UtMDdmMGJmNDhlZDU1XkEyXkFqcGc@._V1_FMjpg_UX1000_.jpg"
    />
    <div
      class="movie-tile-info"
    >
      <h3>
        Mission Impossible
      </h3>
      <div
        class="movie-tile-info-year"
      >
        <h2>
          2018
        </h2>
      </div>
    </div>
    <p>
      Drama,Thriller
    </p>
  </div>
</DocumentFragment>
`;

exports[`MovieTile component tests renders snapshot 1`] = `
<DocumentFragment>
  <div
    class="movie-tile"
    data-testid="info"
  >
    <img
      alt="Mission Impossible"
      src="https://m.media-amazon.com/images/M/MV5BN2U4OTdmM2QtZTkxYy00ZmQyLTg2N2UtMDdmMGJmNDhlZDU1XkEyXkFqcGc@._V1_FMjpg_UX1000_.jpg"
    />
    <div
      class="movie-tile-info"
    >
      <h3>
        Mission Impossible
      </h3>
      <div
        class="movie-tile-info-year"
      >
        <h2>
          2018
        </h2>
      </div>
    </div>
    <p>
      Drama,Thriller
    </p>
  </div>
</DocumentFragment>
`;
